File: Trees/BinaryTreeJava/src/binarytree/BottomView.java
package binarytree;

import java.util.*;

public class BottomView {
    public static void main(String[] args) {
        Node root = null;
        DiameterOrWidth.Height h = new DiameterOrWidth.Height();
        BinaryTree s = new BinaryTree();
        root = s.buildTree(root);
        ArrayList<Integer> ans  = bottomView(root);
        System.out.println(ans);
        //1 2 4 -1 -1 5 -1 -1 3 6 -1 -1 -1
    }
    static class Pair{
        Node node;
        int hd;
        Pair(Node node , int hd){
            this.node = node ;
            this.hd = hd;
        }
    }
    public static ArrayList<Integer> bottomView(Node root){
        Map<Integer,Integer > map = new TreeMap<>();
        ArrayList<Integer> ans = new ArrayList<>();
        if(root == null){
            return ans ;
        }
        Queue<Pair> q = new LinkedList<>();
        q.add(new Pair(root,0));

        while(!q.isEmpty()){
            Pair temp = q.remove();
            Node front = temp.node;
            int hd = temp.hd;
            map.put(hd,front.data);

            if(front.left != null ){
                q.add(new Pair(front.left,hd-1));
            }
            if(front.right != null ){
                q.add(new Pair(front.right,hd+1));
            }
        }
        for (Map.Entry<Integer, Integer> i : map.entrySet()) {
            ans.add(i.getValue());
        }
        return ans ;
    }
}

File: Trees/BinaryTreesC++/bottomView.cpp
#include <bits/stdc++.h>
using namespace std;

class Node{
	public:
		int data;
		Node* left;
		Node* right;
		Node(int d){
			this->data = d;
			this->left = NULL;
			this->right = NULL;
		}
};
Node* buildTree(Node* root){	
	int data; 
	cin>>data ;
	root = new Node(data);
	if(data == -1){
		return NULL;
	}
	

	root->left = buildTree(root->left);
	root->right = buildTree(root->right);
	return root;
}

vector<int> bottomView(Node* root){
	map<int, int> nodes ;
	vector<int> ans ;
	if(root == NULL){
		return ans ;
	}
	queue<pair<Node* , int>> q ;
	q.push(make_pair(root,0));

	while(!q.empty()){
		pair<Node* , int> temp = q.front();
		q.pop();
		Node* front =  temp.first;
		int hd = temp.second;
		
		nodes[hd] = front->data;
		

		if(front->left != NULL ){
			q.push(make_pair(front->left , hd - 1));
		}

		if(front->right != NULL){
			q.push(make_pair(front->right, hd+1));
		}
	}
	for(auto i : nodes){
		ans.push_back(i.second);
	}
}

int main(){
	Node* root = NULL ;
	root = buildTree(root);
	vector<int> ans  = bottomView(root);
	for(auto i : ans){
		cout<< i <<" ";
	}
}
